src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Services/ClientServices.cs

```diff
@@ -24,32 +25,217 @@ namespace Microsoft.Azure.IIoT.OpcUa.Protocol.Services {
     /// Opc ua stack based service client
     /// </summary>
     public class ClientServices : IClientHost, IEndpointServices, IEndpointDiscovery,
-        IDisposable {
+        IDisposable{
 
         /// <inheritdoc/>
-        public X509Certificate2 Certificate { get; private set; }
+        public X509Certificate2 Certificate {
+
+            get {
+
+                return _OpcApplicationConfig.SecurityConfiguration.ApplicationCertificate.Certificate;
+            }
+            private set {
+
+                if (value == null) {
+                    _logger.Warning("Certificate setter called with empty certificate");
+                    return;
+                }
+                if (!value.HasPrivateKey) {
+
+                    //TODO throw exception
+                    //  invalid certificate provided
+                    _logger.Warning("Certificate setter called with a certificate without private key");
+                    return;
+                }
+               
+                _OpcApplicationConfig.SecurityConfiguration.ApplicationCertificate.Certificate = value;
+                _OpcApplicationConfig.CertificateValidator.UpdateCertificate(_OpcApplicationConfig.SecurityConfiguration);
+
+                // copy the certificate into the trusted certificates list
+                try {
+
+                    // add new certificate.
+                    X509Certificate2 publicKey = new X509Certificate2(value.RawData);
+                                     
+                    ICertificateStore trustedStore = _OpcApplicationConfig.SecurityConfiguration.TrustedPeerCertificates.OpenStore();
+
+                    try {
+                        _logger.Information($"Adding own certificate in the certificate trusted peer store. " +
```

Comment:
Structured logging is nice, you should make sure to use it and not inline the string.   That way the additional information can be compressed by serilog.   E.g. in this case StorePath={path]", path, ...  - see other examples in the code base.

Reply:
done, hopefully reasonably comprehensive, please review